src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -666,6 +666,17 @@ public override BoundNode DefaultVisit(BoundNode node)
 
         #region Expression Visitors
 
+        public override BoundNode VisitLoweredIsPatternExpression(BoundLoweredIsPatternExpression node)
+        {
+            var builder = new BoundSpillSequenceBuilder(node.Syntax);
+            builder.AddLocals(node.Locals);
+            builder.AddStatements(VisitList(node.Statements));
+            var temp = Spill(builder, node.Update(
+                ImmutableArray<LocalSymbol>.Empty, ImmutableArray<BoundStatement>.Empty,
+                node.WhenTrueLabel, node.WhenFalseLabel, VisitType(node.Type)));
+            return builder.Update(temp);
```

Comment:
I already mentioned that several times: "The BoundLoweredIsPatternExpression node shouldn't be spilled by default". Also, I believe I already mentioned, that I think that spilling of BoundLoweredIsPatternExpression should be made by value. This is just a regular value, we shouldn't need any new code to spill it. I expect very little changes for the new node in this component. Pretty much all we need to worry about here is that VisitLoweredIsPatternExpression does the right thing when the input value triggers spilling itself (for example it is a method call with await inside). Even in this case, we should look for a way to rewrite the node in such a way that BoundLoweredIsPatternExpression survives, not in the original form, of course. In other words, ideally the value should still be represented as a BoundLoweredIsPatternExpression.

> otherwise locals are not correctly substituted

If that is what the problem is, then that what we should be addressing. SpillSequenceSpiller is not a solution for every problem. If I understand the problem correctly, SpillSequenceSpiller has very little to do with that. We are adding a new expression node that has locals, like BoundSequence does. Look for places where we do something special about Locals in BoundSequence, and see if we should do something similar. And definitely check all overrides of VisitSequence. For example, MethodToClassRewriter has the following:
```
        public override BoundNode VisitSequence(BoundSequence node)
        {
            var newLocals = RewriteLocals(node.Locals);
            var newSideEffects = VisitList<BoundExpression>(node.SideEffects);
            var newValue = (BoundExpression)this.Visit(node.Value);
            var newType = this.VisitType(node.Type);
            return node.Update(newLocals, newSideEffects, newValue, newType);
        }
```

Reply:
Thanks for the pointer. To simplify, I've removed the locals from the node entirely and used Sequence instead, this does not affect anything semantically, the alternative is to add handling of these locals everywhere which doesn't seem to be helpful. On spilling by value, I did get your point, indeed we should only worry about the input expression and NOT make a temp for the boolean result like that. I think that means we have to have the input expression as a field in the lowered node, however, this would raise some complications because of how it's being evaluated (or not),

- if the input is a tuple, we access each element directly or through a temp per each element if side-effecting
- otherwise, the input is accessed directly or through a temp if side-effecting

Right now, these temp assignments are simply added to the list of statements.. so we'll need to extract all that to an *array* of expressions (in case the input is a tuple) and visit each individually. This makes spiller relatively simpler, but local rewriter and possibly emitter will have to be more involved. Again, it's completely possible that I missed something here and there's a better way.

Reply:
Thanks for the pointer. To simplify, I've removed the locals from the node entirely and used Sequence instead, this does not affect anything semantically, the alternative is to add handling of these locals everywhere which doesn't seem to be helpful. On spilling by value, I did get your point, indeed we should only worry about the input expression and NOT make a temp for the boolean result like that. Even now the inner statements are still left with original locals and the substitution seems to be skipped. Any ideas why is that?